// easygit/core.js
var log = require('./log.js');
var funGetInput = require('./get-shell-input');
var Callback = require('./one-by-one');

function funCreateGit(type, strBranchDic, group){
  // 输出错误信息
  function showError(stdout){
    log.main('---------');
    stdout.trim().split(/\n/).forEach(function(str){
      if(str[0] == '#'){
        // log.main(str);
      }else{
        log.alert(str);
      }
    });
    log.main('---------');
    showBranch();
  }
  // 打印当前本地分支名
  function showBranch(){
    log.out('当前分支:' + str_nowBarnch);
  }
  // 根据本地分支名称，转换成远程分支名称
  function getRemoteBranch(str_nowBarnch){
    var arr_remoteName = str_nowBarnch.split('_');
    switch(arr_remoteName[0]){
      case 'la':
      case 'l':
        arr_remoteName[0] = 'fa';
        break;
      case 'ls':
        arr_remoteName[0] = 'fs';
        break;
      case 'b':
      case 'ba':
        arr_remoteName[0] = 'ha';
        break;
      case 'bs':
        arr_remoteName[0] = 'hs';
        break;
    }
    arr_remoteName[0] = arr_remoteName[0] + (group ? '_'+group : '');
    var str_remoteName = arr_remoteName.join('_');
    return str_remoteName;
  }
  var str_nowBarnch = ''; // 本地分支名
  var str_remoteBranch = ''; // 远程分支名
  var bol_hasRemoteBranch = false; // 是否存在远程分支
  var str_commit = ''; // 提交代码说明
  var str_newBranch = ''; // 要创建的分支名，仅在create中使用

  // 获取分支名称，通用
  var obj_getBranch = {
    commend: 'git branch -a',
    callback: function(err, stdout, stderr){
      var arr_branch = stdout.split('\n');
      var arr_nowBranch = arr_branch.filter(function(branch){
        return branch[0] == '*';
      });
      str_nowBarnch = arr_nowBranch[0].substr(2);
      str_remoteBranch = getRemoteBranch(str_nowBarnch);
      var arr_remoteBranch = arr_branch.filter(function(branch){
        return new RegExp('([\\s|\\/])'+str_remoteBranch+'$').test(branch);
        // return (branch.lastIndexOf(str_remoteBranch) != -1);
      });
      if(arr_remoteBranch.length) bol_hasRemoteBranch = true;
      this.fire();
    }
  };

  var obj_gitType = {
    update : [
      obj_getBranch,
      {
        commend : 'git remote update',
        callback : function(err, stdout, stderr){
          if(err){
            log.alert('remote update 失败');
            showError(stdout);
            log.out('操作代码：git remote update');
            return;
          }
          log.out('remote update 成功');
          this.fire();
        }
      },
      {
        commend : 'git merge --no-ff origin master',
        callback : function(err, stdout, stderr){
          if(err){
            log.alert('merge master 失败');
            showError(stdout);
            log.out('操作代码：git merge --no-ff origin master');
            return;
          }
          log.out('merge master 成功');
          log.out('----------');
          log.out(stdout);
          if(bol_hasRemoteBranch){
            this.fire();
          }else{
            this.end();
          }
        }
      },
      {
        commend : function(){
          return 'git merge --no-ff '+strBranchDic+'/'+str_remoteBranch;
        },
        callback : function(err, stdout, stderr){
          if(err){
            log.alert('merge '+str_remoteBranch+' 失败，可能是修改了相同的代码，请检查');
            showError(stdout);
            showError(stderr);
            log.out('操作代码：git merge --no-ff '+strBranchDic+'/'+str_remoteBranch);
            return;
          }
          log.out('merge '+str_remoteBranch+' 成功');
          log.out('----------');
          log.out(stdout);
          this.fire();
        }
      }
    ],
    push : [
      obj_getBranch,
      {
        commend : function(){
          return 'git push '+strBranchDic+' '+str_nowBarnch+':'+str_remoteBranch;
        },
        callback : function(err, stdout, stderr){
          if(err){
            log.alert('push '+str_remoteBranch+' 失败');
            showError(stderr);
            log.out('操作代码：git push '+strBranchDic+' '+str_nowBarnch+':'+str_remoteBranch);
            return;
          }
          log.out('push '+str_remoteBranch+' 成功');
          this.fire();
        }
      }
    ],
    commit : [
      obj_getBranch,
      {
        commend: function(){
          return 'git add -A';
        },
        callback: function(err, stdout, stderr){
          if(err){
            log.alert('git add 出错');
            showError(stdout);
            log.out('操作代码：git add -A');
            return;
          }
          log.out('git add 成功');
          this.fire();
        },
        wait: function(callback){
          log.out('提交代码描述');
          funGetInput(function(input){
            str_commit = input;
            callback();
          });
        }
      },
      {
        commend: function(){
          return 'git commit -m "' + str_commit + '"';
        },
        callback: function(err, stdout, stderr){
          if(err){
            log.alert('本地提交代码出错');
            showError(stdout);
            log.out('操作代码：git commit -m "' + str_commit + '"');
            return;
          }
          log.out('本地提交代码成功');
          this.fire();
        }
      }
    ],
    create : [
      obj_getBranch,
      // 检查当前分支是否有为提交代码
      {
        commend: 'git diff',
        callback: function(err, stdout, stderr){
          if(stdout.trim().length){
            log.alert('本分支还有未提交的代码！');
            this.end();
            return;
          }
          this.fire();
        }
      },
      // 查看是否存在要创建的分支，存在就暂停
      {
        commend: 'git branch -a',
        callback: function(err, stdout, stderr){
          var arr_branch = stdout.split('\n');
          var arr_newBranch = arr_branch.filter(function(branch){
            return new RegExp('([\\s|\\/])'+str_newBranch+'$').test(branch);
          });

          if(arr_newBranch.length){
            log.alert('存在要创建的分支');
            return;
          }
          this.fire();
        },
        wait: function(callback){
          log.out('输入要创建的分支名：');
          funGetInput(function(input){
            str_newBranch = input;
            str_nowBarnch = input;
            callback();
          });
        }
      },
      // 切换到主分支，准备更新
      {
        commend: 'git checkout master',
        callback: function(err, stdout, stderr){
          if(err){
            log.alert('切换分支出错');
            showError(stdout);
            log.out('操作代码：git checkout master');
            return;
          }
          log.out('切换到master分支');
          this.fire();
        }
      },
      // 更新主分支
      {
        commend: 'git pull',
        callback: function(err, stdout, stderr){
          if(err){
            log.alert('拉取代吗失败');
            showError(stdout);
            log.out('操作代码：git pull');
            return;
          }
          log.out('master分支以更新到最新');
          this.fire();
        }
      },
      {
        commend: function(){
          return 'git checkout -b '+str_newBranch;
        },
        callback: function(err, stdout, stderr){
          if(err){
            log.alert('创建分支失败');
            showError(stdout);
            log.out('操作代码：git checkout -b '+str_newBranch);
            return;
          }
          log.out('已创建分支 '+str_newBranch);
          this.fire();
        }
      }
    ],
    PUSH : [
      {
        commend : 'git remote update',
        callback : function(err, stdout, stderr){
          if(err){
            log.alert('remote update 失败');
            showError(stdout);
            log.out('操作代码：git remote update');
            return;
          }
          log.out('remote update 成功');
          this.fire();
        }
      },
      // 查看是否存在要创建的分支，存在就暂停
      {
        commend: 'git branch -a',
        callback: function(err, stdout, stderr){
          var arr_branch = stdout.split('\n');
          var arr_remoteBranch = arr_branch.filter(function(branch){
            return new RegExp('([\\s|\\/])'+str_remoteBranch+'$').test(branch);
          });

          if(!arr_remoteBranch.length){
            log.alert('不存在远程分支');
            return;
          }
          this.fire();
        },
        wait: function(callback){
          log.out('输入要合并的远程分支名：');
          funGetInput(function(input){
            str_remoteBranch = input;
            callback();
          });
        }
      },
      // 切换到主分支，准备创建分支
      {
        commend: 'git checkout master',
        callback: function(err, stdout, stderr){
          if(err){
            log.alert('切换分支出错');
            showError(stdout);
            log.out('操作代码：git checkout master');
            return;
          }
          log.out('切换到master分支');
          this.fire();
        }
      },
      // 更新主分支
      {
        commend: 'git pull',
        callback: function(err, stdout, stderr){
          if(err){
            log.alert('拉取代吗失败');
            showError(stdout);
            log.out('操作代码：git pull');
            return;
          }
          log.out('master分支以更新到最新');
          this.fire();
        }
      },
      // 删除merge分支
      {
        commend: 'git branch -D m',
        callback: function(err, stdout, stderr){
          log.out('删除分支m');
          this.fire();
        }
      },
      // 创建merge分支
      {
        commend: 'git checkout -b m',
        callback: function(err, stdout, stderr){
          if(err){
            log.alert('创建分支m失败');
            showError(stdout);
            log.out('操作代码：git checkout -b m');
            return;
          }
          log.out('创建分支m');
          this.fire();
        }
      },
      // 更新远程分支到本地
      {
        commend : function(){
          return 'git merge --no-ff '+strBranchDic+'/'+str_remoteBranch;
        },
        callback : function(err, stdout, stderr){
          if(err){
            log.alert('merge '+str_remoteBranch+' 失败，可能是修改了相同的代码，请检查');
            showError(stdout);
            showError(stderr);
            log.out('操作代码：git merge --no-ff '+strBranchDic+'/'+str_remoteBranch);
            return;
          }
          log.out('merge '+str_remoteBranch+' 成功');
          log.out('----------');
          log.out(stdout);
          this.fire();
        }
      },
      // 切换到主分支，准备更新
      {
        commend: 'git checkout master',
        callback: function(err, stdout, stderr){
          if(err){
            log.alert('切换分支出错');
            showError(stdout);
            log.out('操作代码：git checkout master');
            return;
          }
          log.out('切换到master分支');
          this.fire();
        }
      },
      // 更新merge分支到master
      {
        commend : 'git merge --no-ff m',
        callback : function(err, stdout, stderr){
          if(err){
            log.alert('merge '+str_remoteBranch+' 失败，可能是修改了相同的代码，请检查');
            showError(stdout);
            showError(stderr);
            log.out('操作代码：git merge --no-ff '+strBranchDic+'/'+str_remoteBranch);
            return;
          }
          log.out('merge '+str_remoteBranch+' 成功');
          log.out('----------');
          log.out(stdout);
          log.out('都完事儿了，还得手动打tag');
          this.fire();
        }
      }
    ]

  };
  return new Callback(obj_gitType[type], showBranch);
}
module.exports = funCreateGit;
